### ISSUE TITLE
[DX] Exception thrown in a sub-controller is hard to debug

### ISSUE THREAD

**Issue by sustmi** [twigphp/Twig#2000]
## TL;DR

Exception thrown in sub-controller shows you a stack trace of Twig internals but nothing about where the original exception was thrown.
## The whole story

When I call `throw new \MyConcreteException('My concrete exception message')` somewhere in a sub-controller action, `Template` class catches the exception and [wraps it in `Twig_Error_Runtime`](https://github.com/twigphp/Twig/blob/1.x/lib/Twig/Template.php#L182).
The error message then looks like this:

```
An exception has been thrown during the rendering of a template ("My concrete exception message") in MyBundle:MyController:myTemplate.html.twig at line 4. 
```

where `myTemplate.html.twig` looks like:

``` lang=twig
{% extends 'MyBundle::base.html.twig' %}

{% block mainContent %}
    {{ render(controller('MyBundle:MyController:someAction')) }}
...
```

The problem is that the information that the exception was thrown in myTemplate.html.twig at line 4 does not help me much to find the cause of the problem.
The stack trace of the exception looks like this:
![exception](https://cloud.githubusercontent.com/assets/885946/13556649/5fa76c14-e3df-11e5-9218-f98bfe3d6c60.png)
As you can see this also tells me nothing about where the `MyConcreteException` was thrown.
I had to scroll down and unfold the previous exception. It took me a while until I discovered how to get the original stack trace.
(Sometimes I forget about this "hidden" backtrace and spend minutes trying to backtrace the exception by myself because I think that there is no other option.)
## What can we do?

I tried changing the exception message to something like this:
`An exception has been thrown during the rendering of a template ("%s"). See the previous exception for details.`

But then [Error::updateRepr() method](https://github.com/twigphp/Twig/blob/1.x/lib/Twig/Error.php#L183) adds information about template file so the final message looks like:

```
An exception has been thrown during the rendering of a template ("My concrete exception message"). See the previous exception for details in MyBundle:MyController:myTemplate.html.twig at line 4
```

which is also confusing.

This is the best form I was able to come with:

```
An exception (see the previous exception for details) has been thrown during the rendering of a template (message: "My concrete exception message") in MyBundle:myController:myTemplate.html.twig at line 4.
```

Btw: Is it necessary to print the original error message in parentheses? At all places I can think of (default debugging error page in Symfony, Symfony Console, Monolog) you can see the wrapping exception as well as the wrapped one.

**Comment by sustmi** [comment]
Also note that the code is duplicated so the exception message should be changed at two places: https://github.com/twigphp/Twig/blob/1.x/lib/Twig/Template.php#L182
https://github.com/twigphp/Twig/blob/1.x/lib/Twig/Template.php#L396

**Comment by sustmi** [comment]
Another variant:
```
An exception has been thrown during the rendering of a template. See the previous exception for details. Original message: "My concrete exception message" in MyBundle:MyController:myTemplate.html.twig at line 4. 
```

**Comment by iltar** [comment]
I recently had a discussion with a colleague if there was anything we could do about this message. Every single day we 1 different person watching the logs etc to see if something is broken and needs fixing.

One of the complaints was the "cryptic" error when you have an error during a sub-request as mentioned here. I have no idea how to fix it properly, but I would love to see it solved as it complicates searching.

**Comment by sustmi** [comment]
I guess that we cannot get rid of wrapping the exception so there will always one more exception  stacked in front of the interesting one.
Would the "See the previous exception for details." note in the wrapping exception "decrypt" it for you, @iltar ?

**Comment by sustmi** [comment]
I guess that we cannot get rid of wrapping the exception so there will always be one more exception  stacked in front of the interesting one.
Would the "See the previous exception for details." note in the wrapping exception "decrypt" it for you, @ilt

**Comment by iltar** [comment]
The conclusion at my company was that the top level exception is useless because it doesn't give us any useful information. Sure it happened during the rendering of a template, but maybe the sub-request part shouldn't be called "during rendering" as it was processing a controller most likely.

**Comment by sustmi** [comment]
Do you think this message will be better to understand, @iltar ?
> An exception has been thrown in a code run from a template. See the previous exception for details. Original message: "My concrete exception message" in MyBundle:MyController:myTemplate.html.twig at line 4.

**Comment by sustmi** [comment]
Btw, the top level message is completely not useless. It gives you information about the template context where the error happened, so that you know which template was the core run throught.

**Comment by stof** [comment]
> Sure it happened during the rendering of a template, but maybe the sub-request part shouldn't be called "during rendering" as it was processing a controller most likely.

Well, Twig has no idea that a sub request was handled by the kernel as part of the template rendering (it does not even know what the kernel is)

**Comment by fabpot** [comment]
Closing this one as it looks like more related to Symfony than Twig.

**Comment by stof** [comment]
@fabpot this is partially related to Twig, as Twig wraps everything into a Twig_Error_Runtime
